rwkv_graph.inc

```diff
@@ -320,6 +303,185 @@ static struct ggml_tensor * rwkv_att_v5(
+    );
+
+    xxx = ggml_cont_inplace(ctx, ggml_permute(ctx, xxx, 3, 2, 0, 1)); // !!!!!
+
+    // xxx = torch.bmm(xxx, tm_w2).view(5, T, -1)
+    // xxx (32, 1, 5, T) @ att_time_maa_w2 (32, 2048, 5) -> (1, 2048, 5, T) -> (5, 2048, T, 1) -> xxx (5, 2048, T)
+    xxx = ggml_permute(ctx, ggml_mul_mat(ctx, xxx, layer.att_time_maa_w2), 3, 1, 0, 2); // !!!!!
+    xxx = ggml_cont_inplace(ctx, xxx); // !!!!!!
+    xxx = ggml_reshape_3d(
+        ctx,
+        xxx,
+        5, layer.att_time_maa_w2->ne[1], sequence_length
+    );
+
+    // mw (2048, T) * att_time_maa_w (2048) + sx (2048, T) -> xw (2048, T)
+    // !!!! 检查是否正确，因为转置
+    struct ggml_tensor * mw = ggml_view_1d(ctx, xxx, n_embed, n_embed * 0 * sizeof(float));
+    struct ggml_tensor * mk = ggml_view_1d(ctx, xxx, n_embed, n_embed * 1 * sizeof(float));
+    struct ggml_tensor * mv = ggml_view_1d(ctx, xxx, n_embed, n_embed * 2 * sizeof(float));
+    struct ggml_tensor * mr = ggml_view_1d(ctx, xxx, n_embed, n_embed * 3 * sizeof(float));
+    struct ggml_tensor * mg = ggml_view_1d(ctx, xxx, n_embed, n_embed * 4 * sizeof(float));
+
+    struct ggml_tensor * xw = ggml_add_inplace(
+        ctx,
+        ggml_mul_inplace(
+            ctx,
+            ggml_add_inplace(ctx, mw, layer.att_time_maa_w),
+            x_prev
+        ),
+        x
+    );
+
+    struct ggml_tensor * xk = ggml_add_inplace(
+        ctx,
+        ggml_mul_inplace(
+            ctx,
+            ggml_add_inplace(ctx, mk, layer.att_time_maa_k),
+            x_prev
+        ),
+        x
+    );
+
+    struct ggml_tensor * xv = ggml_add_inplace(
+        ctx,
+        ggml_mul_inplace(
+            ctx,
+            ggml_add_inplace(ctx, mw, layer.att_time_maa_v),
+            x_prev
+        ),
+        x
+    );
+
+    struct ggml_tensor * xr = ggml_add_inplace(
+        ctx,
+        ggml_mul_inplace(
+            ctx,
+            ggml_add_inplace(ctx, mr, layer.att_time_maa_r),
+            x_prev
+        ),
+        x
+    );
+
+    struct ggml_tensor * xg = ggml_add_inplace(
+        ctx,
+        ggml_mul_inplace(
+            ctx,
+            ggml_add_inplace(ctx, mg, layer.att_time_maa_g),
+            x_prev
+        ),
+        x
+    );
+
+
+    // w = (time_decay + (torch.tanh(xw @ td_w1) @ td_w2)).view(H, S, 1)
+    // att_time_decay_w1 (2048, 64) @ xw (2048, T) -> (64, T)
+    // att_time_decay_w2 (64, 2048) @ (64, T) -> (2048, T) -> w (1, S, H, T) <=> (1, S, H, T) att_heads
+    struct ggml_tensor * w = ggml_add_inplace(
+        ctx,
+        ggml_mul_mat(
+            ctx,
+            layer.att_time_decay_w2,
+            ggml_tanh_inplace(
+                ctx,
+                ggml_mul_mat(ctx,layer.att_time_decay_w1, xw)
+            )
+        ),
+        layer.att_time_decay
+    );
+    w = ggml_reshape_4d(ctx, w, 1, head_size, head_count, sequence_length);
+
+    // w = torch.exp(-torch.exp(w))
+    w = rwkv_exp(ctx,rwkv_1_minus_x(ctx,rwkv_exp(ctx, w)));
```

Comment:
Does this call to rwkv_1_minus_x() belong here? The python function isn't `w = torch.exp(1 - torch.exp(w))`, so shouldn't it be `w = rwkv_exp (ctx, ggml_mul (ctx, -1, rwkv_exp (ctx, w)));`

Reply:
`rwkv_1_minus_x` is defined as `1 - x`, `ggml_neg` is defined as `-x`. So the code looks correct to me -- indeed, we want `exp(1 - exp(w))`, not `exp(-exp(w))`.

Reply:
This refers to the line here: https://github.com/BlinkDL/ChatRWKV/blob/8c7956743703afddd9bbb09ec5fbaf95e5b05227/RWKV_v6_demo.py#L187 

- `w = torch.exp(-torch.exp(w.float()))`

There's no subtraction operation, only negation

Reply:
Oh, okay. I misread your comment, sorry; looks like you are right :)